src/error/error.h

```diff
@@ -0,0 +1,189 @@
+#pragma once
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+// Error utilities.
+
+/// Base error interface.
+/// All error classes must derive from this one.
+class Error {
+ public:
+  /// Conversion to a string for display.
+  virtual std::string to_string() const = 0;
+  virtual ~Error() = default;
+};
+
+/// Basic error, with an error message.
+class GenericError : public Error {
+ public:
+  /// Build a new generic error, with an error message.
+  explicit GenericError(std::string message)
+      : Error(), message_(std::move(message)) {}
+
+  /// Return the error message it was constructed with.
+  std::string to_string() const override { return message(); }
+
+  ~GenericError() override = default;
+
+ protected:
+  /// Access to the field containing the error message passed to the
+  /// constructor.
+  std::string message() const { return message_; }
+
+ private:
+  const std::string message_;
+};
+
+/// Represents either an error or a T.
+/// This class is immutable (but the T can be mutable).
+template <typename T, typename Err = GenericError>
+class ErrorOr {
+  static_assert(std::is_base_of<Error, Err>::value,
+                "Error type must be a subclass of Error");
+  template <typename E>
+  using enable_if_error = typename std::enable_if<std::is_base_of<
+      typename std::decay<Err>::type, typename std::decay<E>::type>::value>;
+
+ public:
+  // Constructors from value or errors.
+
+  // Not explicit on purpose.
+  template <typename E>
+  ErrorOr(E error,  // NOLINT: explicit
+          typename enable_if_error<E>::type* /*unused*/ = nullptr)
+      : union_(std::move(error)), is_ok_(false) {}
+
+  // Not explicit on purpose.
+  ErrorOr(T value)  // NOLINT: explicit
+      : union_(std::move(value)),
+        is_ok_(true) {}
+
+  // Move constructor.
+  template <typename E>
+  ErrorOr(ErrorOr<T, E>&& other)  // NOLINT: explicit
+      : is_ok_(other.is_ok_) {
+    if (is_ok_)
+      new (&union_.value) T(std::move(other.union_.value));
+    else
+      new (&union_.error) std::unique_ptr<Err>(std::move(other.union_.error));
+  }
+
+  // Move assignment.
+  template <typename E>
+  ErrorOr& operator=(ErrorOr<T, E>&& other) {
+    this->~ErrorOr();
+    is_ok_ = other.is_ok_;
+    if (is_ok_)
+      new (&union_.value) T(std::move(other.union_.value));
+    else
+      new (&union_.error) std::unique_ptr<Err>(std::move(other.union_.error));
+    return *this;
+  }
+
+  // Copy constructor.
+  ErrorOr(const ErrorOr& other) = delete;
+  // Copy assignment.
+  ErrorOr& operator=(const ErrorOr& other) = delete;
+
+  /// Check whether it is an error or a value.
+  bool is_ok() const { return is_ok_; }
+
+  /// Return the value if it is one, fail otherwise.
+  T& value_or_die() {
```

Comment:
This function is not const for people who want to get the value in the ErrorOr and modify it. There is no error_or_die that is not const, because there is no use case (so far) for modifying the error.